### PR TITLE
feat(reload): per-archive dynamic reload with shared cache

### PR DESCRIPTION
**Pull request by hooyao**
## Summary

- Replace experimental VfsScope "nuke everything" reload with surgical per-archive add/remove
- FileSystemWatcher monitors archive directory; ArchiveChangeConsolidator batches events into net deltas
- Per-archive ArchiveNode ref counting with drain-before-removal ensures clean concurrent shutdown
- Shared cache infrastructure stays alive — only affected archive's data is cleaned up
- 36 new tests, 382 total passing

## Key Components

| Component | Purpose |
|-----------|---------|
| `GenericCache.TryRemove` | Targeted entry removal with orphan tracking |
| `FileContentCache.RemoveArchive` | Per-archive key index for efficient cleanup |
| `ArchiveTrie` (RWLock) | Thread-safe concurrent read/write |
| `ArchiveNode` / `ArchiveGuard` | Per-archive drain with TryEnter/Exit pattern |
| `IArchiveManager` | ISP — archive lifecycle separate from VFS ops |
| `ArchiveChangeConsolidator` | Event batching with state machine (add+delete=noop) |
| `DokanHostedService` | FileSystemWatcher + filtering + retry probe |
| `DynamicReloadSuite` | Endurance test with 5 concurrent workloads |

## Design Doc

Full design at `src/Docs/DYNAMIC_RELOAD_DESIGN.md` — reviewed by 5 specialized agents (concurrency, cache architecture, VFS/domain, test coverage, Windows edge cases).

## Test plan

- [ ] 382 non-endurance tests pass (99 domain + 166 caching + 10 filesystem + 74 integration + 33 zip)
- [ ] DynamicReloadSuite endurance test exercises add/remove churn with concurrent reads
- [ ] Verify no regressions in existing endurance test suites
- [ ] Manual test: copy/delete ZIP files while drive is mounted, verify files appear/disappear

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/ZipDrive.Application/Services/ArchiveNode.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    public bool IsDraining => _draining;
+    public int ActiveOps => Volatile.Read(ref _activeOps);
+
+    /// <summary>
+    /// Cancellation token that is cancelled when drain starts.
+    /// Pass this to fire-and-forget operations (e.g., prefetch) so they abort promptly.
+    /// </summary>
+    public CancellationToken DrainToken => (_drainCts ??= new CancellationTokenSource()).Token;
+
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/ArchiveNode.cs]: Fixed in cdae38c. DrainToken now uses Interlocked.CompareExchange for thread-safe lazy initialization. Losing CTS is disposed immediately.

File: src/ZipDrive.Application/Services/ArchiveNode.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            return;
+        }
+
+        // Create TCS before setting _draining — volatile write to _draining provides
+        // a release fence, ensuring the prior store to _drainTcs is visible to any
+        // thread that observes _draining == true.
+        _drainTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+        _draining = true;
+
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/ArchiveNode.cs:103]: Addressed by design: DrainAsync is called exclusively from RemoveArchiveAsync (serialized by ApplyDeltaAsync). Single-caller assumption documented in class and method comments. Double-drain guard handles accidental re-entry.

File: src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs [hunk 3]
```diff
@@ -96,46 +100,109 @@ public Task UnmountAsync(CancellationToken cancellationToken = default)
...
 
+    // ── IArchiveManager ─────────────────────────────────────────────────
+
+    /// <inheritdoc />
+    public Task AddArchiveAsync(ArchiveDescriptor archive, CancellationToken ct = default)
+    {
+        ArgumentNullException.ThrowIfNull(archive);
+        _archiveTrie.AddArchive(archive);
+        _archiveNodes[archive.VirtualPath] = new ArchiveNode(archive);
+        _logger.LogInformation("Archive added: {VirtualPath}", archive.VirtualPath);
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Application/Services/ZipVirtualFileSystem.cs]: Fixed in cdae38c. AddArchiveAsync now uses GetOrAdd to preserve existing ArchiveNode during in-flight operations. Only RemoveArchiveAsync replaces/disposes nodes.

File: src/ZipDrive.Infrastructure.FileSystem/DokanHostedService.cs [hunk 4]
```diff
@@ -1,185 +1,487 @@
...
+
+            _consolidator = new ArchiveChangeConsolidator(quietPeriod, ApplyDeltaAsync, _logger);
+
+            _watcher = new FileSystemWatcher(_mountSettings.ArchiveDirectory, "*.zip")
+            {
+                NotifyFilter = NotifyFilters.FileName | NotifyFilters.DirectoryName,
+                IncludeSubdirectories = true,
+                InternalBufferSize = 65536, // 64KB — holds ~860 events vs default 8KB
+                EnableRaisingEvents = true
+            };
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.FileSystem/DokanHostedService.cs:195]: Acknowledged limitation. The *.zip filter is intentional for performance — directory renames are caught by the OnWatcherError full reconciliation safety net. Documented in design doc Section 11.1.1.

File: src/ZipDrive.Infrastructure.FileSystem/DokanHostedService.cs [hunk 5]
```diff
@@ -1,185 +1,487 @@
...
+                await _archiveManager.AddArchiveAsync(descriptor);
+                return;
+            }
+
+            if (attempt < delays.Length)
+            {
+                _logger.LogDebug("File not yet readable, retrying in {Delay}s: {Path}",
+                    delays[attempt].TotalSeconds, virtualPath);
+                await Task.Delay(delays[attempt]);
+            }
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.FileSystem/DokanHostedService.cs:375]: Fixed in cdae38c. _stoppingToken stored from ExecuteAsync and passed to Task.Delay in TryAddArchiveAsync. Retry capped at ~10s total.

File: src/ZipDrive.Infrastructure.FileSystem/ArchiveChangeConsolidator.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            _logger.LogError(ex, "Error applying archive change delta");
+        }
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        if (_disposed) return;
+        _disposed = true;
+        await _timer.DisposeAsync();
+    }
```
[the remark this answers is not in the document]
Comment (hooyao) [reply src/ZipDrive.Infrastructure.FileSystem/ArchiveChangeConsolidator.cs:157]: Fixed in cdae38c. _inflightFlush tracks the flush task (assigned before execution starts). DisposeAsync awaits it before returning. Further tightened in 8b2f96e.
